autosklearn/pipeline/components/feature_preprocessing/select_rates_regression.py

```diff
@@ -0,0 +1,126 @@
+from ConfigSpace.configuration_space import ConfigurationSpace
+from ConfigSpace.hyperparameters import UniformFloatHyperparameter, \
+    CategoricalHyperparameter
+
+from autosklearn.pipeline.components.base import \
+    AutoSklearnPreprocessingAlgorithm
+from autosklearn.pipeline.constants import SIGNED_DATA, UNSIGNED_DATA, SPARSE, DENSE, INPUT
+
+
+class SelectRegressionRates(AutoSklearnPreprocessingAlgorithm):
+    def __init__(self, alpha, mode='percentile',
+                 score_func="f_regression", random_state=None):
+        import sklearn.feature_selection
+
+        self.random_state = random_state  # We don't use this
+        self.alpha = alpha
+        self.mode = mode
+
+        if score_func == "f_regression":
+            self.score_func = sklearn.feature_selection.f_regression
+        elif score_func == "mutual_info_regression":
+            self.score_func = sklearn.feature_selection.mutual_info_regression
+            # Work Around as SMAC does not handle Not Equal
+            # Mutual info needs scikit learn default to prevent
+            # running into p_values problem (no pvalue found)
+            self.mode = 'percentile'
+        else:
+            raise ValueError("score_func must be in ('f_regression, 'mutual_info_regression') "
+                             "for task=regression "
+                             "but is: %s " % (score_func))
+
+    def fit(self, X, y):
+        import scipy.sparse
+        import sklearn.feature_selection
+
+        self.alpha = float(self.alpha)
+
+        self.preprocessor = sklearn.feature_selection.GenericUnivariateSelect(
+            score_func=self.score_func, param=self.alpha, mode=self.mode)
+
+        # Because the pipeline guarantees that each feature is positive,
+        # clip all values below zero to zero
+        if self.score_func == sklearn.feature_selection.chi2:
+            if scipy.sparse.issparse(X):
+                X.data[X.data < 0] = 0.0
+            else:
+                X[X < 0] = 0.0
+
+        self.preprocessor.fit(X, y)
+        return self
+
+    def transform(self, X):
+
+        if self.preprocessor is None:
+            raise NotImplementedError()
+        try:
+            Xt = self.preprocessor.transform(X)
+        except ValueError as e:
+            if "zero-size array to reduction operation maximum which has no " \
+                    "identity" in e.message:
+                raise ValueError(
+                    "%s removed all features." % self.__class__.__name__)
+            else:
+                raise e
+
+        if Xt.shape[1] == 0:
+            raise ValueError(
+                "%s removed all features." % self.__class__.__name__)
+        return Xt
+
+    @staticmethod
+    def get_properties(dataset_properties=None):
+        data_type = UNSIGNED_DATA
```

Comment:
I think the data type is only important for classification, where chi2 can only handle positive data.

Reply:
This is a left over from when we had everything in the same file. Fixed.